### ISSUE TITLE
Bayesian Logistic Regression

### ISSUE THREAD

**Issue by lorenzkuhn** [ratschlab/projects2019-boosting-bbvi-pyro#7]
Apply prototype to BLR.

**Comment by gideonite** [comment]
- Make another prototype for this.
- Once we have these two prototypes then we are ready to refactor and make something general.

**Comment by lorenzkuhn** [comment]
Just a quick update on this: I've set up a first version of the Bayesian Logistic Regression model [here](https://github.com/ratschlab/projects2019-boosting-bbvi-pyro/blob/wip/bimodal_experiments/bayesian_logistic_regression.py). I've also started using the ChemReact data with it. There are still to be some bugs that I need to find, as the SVI for the components currently diverges. 

This is the new model: 
https://github.com/ratschlab/projects2019-boosting-bbvi-pyro/blob/1833dbecab22a7ba783625b3c62d9a7186198719/bayesian_logistic_regression.py#L49-L53

and this is the new guide:
https://github.com/ratschlab/projects2019-boosting-bbvi-pyro/blob/1833dbecab22a7ba783625b3c62d9a7186198719/bayesian_logistic_regression.py#L42-L46

As a benchmark, I've also set up a function that allows us to use Hamiltonian Monte Carlo inference to get the weights  - could do the same for regular SVI.

Am I making any obvious mistakes in defining the model and the guide, @gideonite? 
(Not urgent - we can also discuss this tomorrow.)

**Comment by lorenzkuhn** [comment]
@gideonite : Check if you used intercept to fit ChemReact in your Edward implementation

**Comment by lorenzkuhn** [comment]
@lorenzkuhn : Try Laplace guide instead of Gausssian guide

**Comment by lorenzkuhn** [comment]
@lorenzkuhn : Try Gaussian initialisation of guide

**Comment by lorenzkuhn** [comment]
@lorenzkuhn : Try different Adam learning rates.

**Comment by lorenzkuhn** [comment]
@gideonite : Review this: https://github.com/ratschlab/projects2019-boosting-bbvi-pyro/blob/wip/bimodal_experiments/bayesian_logistic_regression.py

**Comment by lorenzkuhn** [comment]
@lorenzkuhn : Apply standard SVI to figure out whether it's the model/guide that's set up incorrectly or the RELBO.

**Comment by lorenzkuhn** [comment]
> @lorenzkuhn : Try different Adam learning rates.

Keeping the other parameters as they are, changing the learning rate does not seem to change the dynamics of the boosting optimisation: The RELBO starts decreasing after a certain number of Adam update steps. The larger the learning rate, the earlier this happens.

**Comment by lorenzkuhn** [comment]
@gideonite : There's no longer a need for you to go through the Bayesian Logistic Regression code, thanks. I found the bug just now and BLR seems to work. Here's the plot for the first component of the approximation, for example: 
![RELBO_fixed](https://user-images.githubusercontent.com/9898136/69535265-d78da600-0f7b-11ea-9e79-af93ac35bb96.png)

The issue was that I had `torch.tensor(input_data)` in my model definition, which creates a copy of the `input_data` tensor which leads to an incorrect computational graph.
